File: src/2.两数相加.ts
/*
 * @lc app=leetcode.cn id=2 lang=typescript
 * https://leetcode.cn/problems/add-two-numbers/description/
 * [2] 两数相加
 */

import { ListNode } from './model/node';

// @lc code=start
function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
  const dummyHead = new ListNode(0);
  let [p1, p2, curr] = [l1, l2, dummyHead];
  // 进位
  let carry = 0;

  while (p1 || p2) {
    // 对于已经遍历到尾部的指针，认为其值为 0
    const sum = (p1 ? p1.val : 0) + (p2 ? p2.val : 0) + carry;
    // 当前位和超过 10，产生进位
    carry = Math.trunc(sum / 10);
    // 生成当前位
    curr.next = new ListNode(sum % 10);
    curr = curr.next;

    // 已经遍历到尾部的指针维持 null 值
    p1 = p1 ? p1.next : null;
    p2 = p2 ? p2.next : null;
  }

  return dummyHead.next;
}
// @lc code=end

File: cpp/src/2.两数相加.cpp
#include <leet/list_node.h>
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */

// @lc code=start
class Solution {
 public:
  ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
    ListNode res, *p = &res;
    int carry = 0;
    while (l1 || l2) {
      // 计算当前位与进位，若其中一条链已经走完了，则认为其当前位为 0
      int sum = (l1 ? l1->val : 0) + (l2 ? l2->val : 0) + carry;
      carry = sum / 10;
      p->next = new ListNode(sum % 10);

      p = p->next;

      // 前往下一位，如果其中一条链已经到尾部了，则维持在尾部
      if (l1) l1 = l1->next;
      if (l2) l2 = l2->next;
    }
    // 考虑最后的进位
    if (carry) {
      p->next = new ListNode(carry);
    }
    return res.next;
  }
};
// @lc code=end
